layabout.py

```diff
@@ -1,3 +1,287 @@
+                handler when the event is triggered.
+
+        Returns:
+            A decorator that validates and registers a Layabout event handler.
+
+        Raises:
+            TypeError: If the decorated :obj:`Callable`'s signature does not
+                accept at least 2 parameters.
+        """
+        def decorator(fn: Callable) -> Callable:
+            # Validate that the wrapped callable is a suitable event handler.
+            sig = signature(fn)
+            if len(sig.parameters) < 2:
+                raise TypeError("Layabout event handlers take at least 2 "
+                                "positional arguments, a slack client and an "
+                                "event")
+
+            # Register a tuple of the callable and its kwargs, if any.
+            self._handlers[type].append((fn, kwargs or {}))
+
+            @wraps(fn)
+            def wrapper(slack: SlackClient, event: Dict[str, Any],
+                        **kwargs: Any) -> Any:
+                # We don't actually do anything with the return value, but this
+                # might make unit testing easier for users.
+                return fn(slack, event, **kwargs)
+            return wrapper
+        return decorator
+
+    def connect(self, token: str = None) -> bool:
+        """
+        Attempt to establish or reset a connection to Slack's API.
+
+        Note:
+            It isn't normally necessary to use this method as
+            :obj:`Layabout.run` will take care of establishing a connection for
+            you. This is mostly helpful if you want to take advantage of the
+            embedded :obj:`slackclient.SlackClient` on the :obj:`Layabout`
+            instance *before* entering into the :obj:`Layabout.run` loop.
+
+        Args:
+            token: A Slack API token. If given it will override an existing
+                connection (if one exists), otherwise an existing token or
+                an environment variable will be used.
+
+        Returns:
+            Whether the connection succeeded.
+
+        Raises:
+            MissingSlackToken: If no API token is available.
+        """
+        resetting = False
+
+        # If we were given a token let's prefer the new one and establish or
+        # reset the connection.
+        if token is not None:
+            self._token = token
+            resetting = True
+
+        # We don't have an existing token, so let's try to get one from an
+        # environment variable or give up.
+        if self._token is None:
+            try:
+                self._token = os.environ[self._env_var]
+            except KeyError:
+                raise MissingSlackToken('Cannot connect to the Slack API'
+                                        ' without a token')
+
+        # Either we've never connected before or we're purposefully resetting
+        # the connection.
+        if self.slack is None or resetting:
```

Comment:
I see the value of the `getToken` function, but I feel like this might be taking it too far. Once you remove the `TODO` below this this is honestly just two lines. ¯\\\_(ツ)_/¯

I guess if you were to take into account features I've discussed about like adding a user agent to the Slack client, then maybe this would be useful. 🤔

Reply:
I like this idea, because naming it in a function would get rid of the need for your comment I think.

>  A comment is an apology for not choosing a more clear name, or a more reasonable set of parameters, or for the failure to use explanatory variables and explanatory functions.

With this, and the above change about adding a `get_token` function, 24 lines of comments, code and whitespace would become 5 lines of code.  Seems like a win to me.